### PR TITLE
Fixes error when randomizing charged molecules

### PR DESCRIPTION
**Pull request by EBjerrum**
Fixes error when randomizing charged molecules as the one below:

from rdkit import Chem
from rdkit.Chem import Randomize

#Charged molecule
mol = Chem.MolFromSmiles('C[N+](C)(C)C')

#Get molblock and randomize atom order
mb = Chem.MolToMolBlock(mol)
rmb = Randomize.RandomizeMolBlock(mb)

#Randomized MolBlock does not contain charge information
#Conversion back to molecule fails
rmol = Chem.MolFromMolBlock(rmb)
[08:36:58] Explicit valence for atom # 0 N, 4, is greater than permitted

### CODE REVIEW

**Comment by gedeck** [comment]
Can you add a test for this? There are currently no tests for this module. You can make a copy of `UnitTestSATIS.py` and modify it (call it `UnitTestRandomize.py`). The file itself contains actually a few tests `CheckCanonicalization` that could form the basis of how a proper UnitTest.

The new UnitTestRandomize needs to be added to `test_list.py` to include it.

File: rdkit/Chem/Randomize.py [hunk 1]
```diff
@@ -43,6 +43,17 @@ def RandomizeMolBlock(molB):
...
     res.append(inL)
     idx += 1
+  #Charges
+  for i in range(idx, len(splitB)):
+    if splitB[i][0:6] == "M  CHG":
+      line = splitB[i]
+      chargeline = line.split()
+      col = line[0:9]
+      for i in range(3,len(chargeline),2):
+        col = col +"%4i%4i"%(order.index(int(chargeline[i])-1)+1,int(chargeline[i+1])+1)
```
Comment (gedeck) [inline rdkit/Chem/Randomize.py]: Please use spaces around operators as in the rest of the code.

If you use this helper function:
```
def grouped(iterable, n):
  return zip(*[iter(iterable)]*n)
```
you can rewrite the code in a form that may be easier to understand:
```
for atom, charge in grouped(chargeline[3:], 2):
  atom = order.index(int(atom) - 1) + 1
  charge = int(charge) + 1
  col += "%4i%4i" % (atom, charge)
```
I don't understand why you add 1 to the charge (or do I missunderstand)? If you write tests, you should also have a molblock for a negatively charged molecule.

Comment (EBjerrum) [inline rdkit/Chem/Randomize.py]: Thank you for reviewing the code. I remember I had to add + 1 to the index due to zero/one counting differences, but it sounds strange to add + 1 to the charge, I'll see if I can add a sensible test.

Comment (EBjerrum) [inline rdkit/Chem/Randomize.py]: I implemented your suggestions and added a unittest. I'm not so used to the collaborative features of github. After pushing to the branch, it seems like the changes and the new files were added to this pull request, so I assume I do not need to make a new pull request?

Comment (gedeck) [inline rdkit/Chem/Randomize.py]: That is how github works. In addition to automatically adjusting the pull-request, it also triggers the checks. If the changes address a part of the code that was previously commented on, the comment is no longer shown. That makes subsequent reviews also easier.

File: rdkit/Chem/UnitTestRandomize.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  return Chem.MolToSmiles(Chem.MolFromMolBlock(mb))
+
+class TestCase(unittest.TestCase):
+
+  def test1(self):
+    smiles = ["C[N+](C)(C)C", "CC(=O)[O-]", "[NH3+]CC(=O)[O-]",'[Cl-].[Cl-].[Mg+2]']
+    msg = "Could not randomize charged mol %s"
+    for smi in smiles:
+      mb = Chem.MolToMolBlock(Chem.MolFromSmiles(smi))
+      rmb = Randomize.RandomizeMolBlock(mb)
```
Comment (gedeck) [inline rdkit/Chem/UnitTestRandomize.py]: While I don't see that it would cause a problem here, the use of random will create a _different_ test every time. You can seed the random module using e.g. `random.seed(0)`. This will create the same randomized mol block every time. 

To cover a few more cases, you could then also repeat the randomization for every molecule several times (10, 20, or so) as it's done in `test2`

It's also useful to test edge cases, MgCl2 is a good one as it only contains charged atoms. [O-][O-] would also be interesting as it will have the charged atom as the first and the last atom; [O-]C[O-] would also be good for the same reason.

File: rdkit/Chem/UnitTestRandomize.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    smiles = ['CON', 'c1ccccn1', 'C/C=C/F']
+    msg = "\nRef: %s\n   : %s"
+    nReps = 10
+    for smi in smiles:
+      mol = Chem.MolFromSmiles(smi)
+      refSmi = Chem.MolToSmiles(mol, False)
+      for i in range(nReps):
+        m2 = Randomize.RandomizeMol(mol)
+        randomizedsmi = Chem.MolToSmiles(m2, False)
+        self.assertEqual(refSmi, randomizedsmi, msg=msg % (refSmi, smi))
```
Comment (gedeck) [inline rdkit/Chem/UnitTestRandomize.py]: This is essentially the `CheckCanonicalization` from the `Randomize.py` file. The function and the `__main__` block can be deleted, however @greglandrum may want to comment here.

File: rdkit/Chem/UnitTestRandomize.py (hunk 2 above)
Comment (EBjerrum) [inline rdkit/Chem/UnitTestRandomize.py]: Yes, its probably best to fix the random seed. I also added '[Na+].[Na+].[O-2]' to have a case with a double negative charge.

File: rdkit/Chem/UnitTestRandomize.py (hunk 3 above)
Comment (EBjerrum) [inline rdkit/Chem/UnitTestRandomize.py]: I was not sure exactly what was being tested there, so I more or less ported it as it was. Seems to me that it more tests the canonization of the SMILES rather than it actually tests the randomization. Maybe that was what the randomization script was there for in the first place??

Comment (EBjerrum) [inline rdkit/Chem/UnitTestRandomize.py]: What is the status of this pull request. I'm not used to github? Is there anything I need to do from my end??

Comment (greglandrum) [inline rdkit/Chem/UnitTestRandomize.py]: It's blocked on me reviewing it. I'll try to do that this weekend. Sorry for the delay

File: rdkit/Chem/UnitTestRandomize.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import unittest
+
+from rdkit import Chem
+from rdkit.Chem import Randomize
+import random
+
+
+def MolBlockToSmiles(mb):
+  return Chem.MolToSmiles(Chem.MolFromMolBlock(mb))
```
Comment (greglandrum) [inline rdkit/Chem/UnitTestRandomize.py]: It's likely not relevant here, but normally you would want to include the `isomericSmiles=True` argument to `Chem.MolToSmiles()` so that you get stereochemistry.

**Comment by greglandrum** [comment]
@EBjerrum I'm cleaning up old PRs. The last comment here was me pointing out a simpler way of solving this problem. Are you willing to revisit this and finish it or should I just close it?

**Comment by EBjerrum** [comment]
Hi Greg, Im out of office the next couple of weeks. Its kind of long time since I looked at it, and personally I dont use it anymore, theres the solution you pointed out and also now the doRandom flag for MolToSmiles, which was my original intent. So just close it.
BR Esben

Skickat från Yahoo Mail på Android 
 
  På fre, juni 2019 klockan 11:31, Greg Landrum<notifications@github.com> skrev:   
@EBjerrum I'm cleaning up old PRs. The last comment here was me pointing out a simpler way of solving this problem. Are you willing to revisit this and finish it or should I just close it?
